File: data_structures_and_algorithms/alvin-zablan/dynamic-programming/part06_canconstruct/javascript/canConstructMemoised.js
const canConstruct = (target, wordBank, memo = {}) => {
  if (target in memo) return memo[target];
  if (target === "") return true;

  for (let word of wordBank) {
    // checking to see if word is a prefix of wordBank
    if (target.indexOf(word) === 0) {
      const suffix = target.slice(word.length);
      if (canConstruct(suffix, wordBank, memo) === true) {
        memo[target] = true;
        return true;
      }
    }
  }

  memo[target] = false;
  return false;
};

// Brute force
// time: O(m^n * m)
// space: O(m^2)

// Memoisation
// time: O(n * m^2)
// space: O(m^2)

console.log(canConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd"])); //true
console.log(canConstruct("scateboard", ["bo", "rd", "ate", "t", "ska", "sk", "boar"])); //false
console.log(canConstruct("enterapotentpot", ["a", "p", "ent", "enter", "ot", "o", "t"])); //true
console.log(canConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef", ["e", "ee", "eee", "eeee", "eeeee", "eeeeee"])); //false

File: data_structures_and_algorithms/alvin-zablan/dynamic-programming/part06_canconstruct/java/CanConstructMemoised.java
package part06_canconstruct.java;

import java.util.HashMap;
import java.util.Map;

public class CanConstructMemoised {
  private static Map<String, Boolean> constructMemo = new HashMap<>();
  public static void main(String[] args) {
    
    System.out.println(canConstruct("abcdef", new String[]{"ab", "abc", "cd", "def", "abcd"}));
    constructMemo.clear();
    System.out.println(canConstruct("skateboard", new String[]{"bo", "rd", "ate", "t", "ska", "sk", "boar"}));
    constructMemo.clear();
    System.out.println(canConstruct("enterapotentpot", new String[]{"a", "p", "ent", "enter", "ot", "o", "t"}));
    constructMemo.clear();
    System.out.println(canConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef", new String[]{"e", "ee", "eee", "eeee", "eeeee", "eeeeee"}));
    constructMemo.clear();
  } 

  public static boolean canConstruct(String targetStr, String[] wordBank){
    if(constructMemo.containsKey(targetStr)){return constructMemo.get(targetStr);}
    if(targetStr == ""){
      return true;
    }

    for(String word : wordBank){
      if(targetStr.indexOf(word) == 0){
        // System.out.println("Works so far..." + word);
        String strRemainder = targetStr.substring(word.length());
        if(canConstruct(strRemainder, wordBank) == true){
          constructMemo.put(strRemainder, true);
          return true;
        }
      }
    }
    constructMemo.put(targetStr, false);
    return false;
  }
}

